generic-sensor/generic-sensor-tests.js

```diff
@@ -0,0 +1,158 @@
+
+  test(() => {
+    let sensor, start_return;
+    sensor = new sensorType();
+    sensor.onerror = () => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    };
+    //The default sensor.state is 'unconnected'
+    assert_equals(sensor.state, "unconnected");
+    start_return = sensor.start();
+    //The sensor.state changes to 'activating' after sensor.start()
+    assert_equals(sensor.state, "activating");
+    //TODO: The permission is not ready.
+    //the sensor.start() return undefined
+    assert_equals(start_return, undefined);
+    sensor.stop();
+  }, "sensor.start() is correct");
+
+  test(() => {
+    let sensor, stop_return;
+    sensor = new sensorType();
+    sensor.onerror = () => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    };
+    sensor.start();
+    stop_return = sensor.stop();
+    //The sensor.state changes to 'idle' after sensor.stop()
+    assert_equals(sensor.state, "idle");
+    //the sensor.stop() returns undefined
+    assert_equals(stop_return, undefined);
+  }, "sensor.stop() is correct");
+
+  async_test(t => {
+    window.onmessage = t.step_func(e => {
+      assert_equals(e.data, "SecurityError");
+      t.done();
+    });
+  }, "throw a 'SecurityError' when firing sensor readings within iframes");
+
+  async_test(t => {
+    let sensor = new sensorType();
+    sensor.onactivate = t.step_func_done(() => {
+      assert_true(verifyReading(sensor));
+      let cachedSensor = sensor;
+      let win = window.open('', '_blank');
+      assert_equals(sensor, cachedSensor);
+      win.close();
+      sensor.stop();
+    });
+    sensor.onerror = t.step_func_done(event => {
+      assert_unreached(event.error.name + ":" + event.error.message);
+    });
+    sensor.start();
+  }, "sensor readings can not be fired on the background tab");
+}
+
+function runGenericSensorInsecureContext(sensorType) {
+  test(() => {
+    assert_throws('SecurityError', () => {
+      let sensor = new sensorType();
+    });
+  }, "throw a 'SecurityError' when construct sensor in an insecure context");
+}
+
+function runGenericSensorOnerror(sensorType) {
+  async_test(t => {
+    let sensor = new sensorType();
+    sensor.onactivate = t.step_func_done(assert_unreached);
+    sensor.onerror = t.step_func_done(event => {
+      assert_equals(sensor.state, 'errored');
+      assert_equals(event.error.name, 'NotReadableError');
+    });
+    sensor.start();
+  }, "'onerror' event is fired when sensor is not supported");
+}
+
+function runSensorFrequency(sensorType) {
+  test(() => {
+    assert_throws(new RangeError(), () => new sensorType({frequency: -60}));
+  }, "negative frequency causes exception from constructor");
+
+  async_test(t => {
+    let sensor = new sensorType({frequency: 600});
+    let number = 0;
+    sensor.onchange = () => {
+      number++;
+    };
+    sensor.onerror = t.step_func_done(event => {
+      assert_unreached(event.error.name + ":" + event.error.message);
```

Comment:
you're using this construct very often. What about extracting it into it's own helper function? E.g.: 

```js
let unreached = event => {
    assert_unreached(event.error.name + ":" + event.error.message);
};

sensor.onerror = t.step_func_done(unreached);
```

Like that if you want to display something different instead, you can do it at once for all cases. Just a thought.